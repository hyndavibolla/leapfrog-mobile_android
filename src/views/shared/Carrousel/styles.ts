import { css } from '@emotion/native';

export const styles = {
  body: css`
    margin-horizontal: -16px;
  `,
  container: css`
    align-items: center;
  `,
  view: css`
    align-items: center;
  `
};
