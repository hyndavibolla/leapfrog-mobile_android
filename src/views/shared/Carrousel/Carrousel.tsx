import React, { memo, ReactElement, useCallback, useRef } from 'react';
import { View, Animated, FlatListProps } from 'react-native';

import { styles } from './styles';

export type Props = {
  children: ReactElement<FlatListProps<any>>;
  separatorWidth: number;
  itemWidth: number;
};

const Carrousel = ({ children, itemWidth, separatorWidth }: Props) => {
  const scrollX = useRef(new Animated.Value(0)).current;
  const { renderItem, contentContainerStyle, style } = children.props;

  const itemSeparatorComponent = useCallback(() => <View style={{ width: separatorWidth, minHeight: 1 }} />, [separatorWidth]);

  return (
    <Animated.FlatList
      {...children.props}
      horizontal
      snapToInterval={itemWidth + separatorWidth}
      decelerationRate={0}
      style={[styles.body, style]}
      showsHorizontalScrollIndicator={false}
      contentContainerStyle={[styles.container, contentContainerStyle]}
      ItemSeparatorComponent={itemSeparatorComponent}
      onScroll={Animated.event([{ nativeEvent: { contentOffset: { x: scrollX } } }], { useNativeDriver: true })}
      renderItem={item => <View style={[styles.view, { width: itemWidth }]}>{renderItem(item)}</View>}
    />
  );
};

export default memo(Carrousel);
