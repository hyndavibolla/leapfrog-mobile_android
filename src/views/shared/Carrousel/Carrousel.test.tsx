import React from 'react';
import { View, Text, FlatList } from 'react-native';
import { render } from '@testing-library/react-native';

import Carrousel from './Carrousel';

const testID = 'carrousel-flat-list';
const data = [].fill('Data', 5).map((item, index) => `${item} ${index}`);
const flatListComponent = (
  <FlatList
    testID={testID}
    data={data}
    renderItem={item => (
      <View>
        <Text>{item}</Text>
      </View>
    )}
  />
);

describe('Carrousel', () => {
  it('should render a Carrousel', () => {
    const { toJSON, queryAllByTestId } = render(<Carrousel children={flatListComponent} separatorWidth={10} itemWidth={90} />);
    expect(toJSON()).toMatchSnapshot();
    expect(queryAllByTestId(testID)).toBeTruthy();
  });
});
