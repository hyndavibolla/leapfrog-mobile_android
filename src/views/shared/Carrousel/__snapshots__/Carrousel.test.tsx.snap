// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carrousel should render a Carrousel 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  collapsable={false}
  contentContainerStyle={
    Array [
      Object {
        "alignItems": "center",
      },
      undefined,
    ]
  }
  data={Array []}
  decelerationRate={0}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  showsHorizontalScrollIndicator={false}
  snapToInterval={100}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "marginHorizontal": -16,
    }
  }
  testID="carrousel-flat-list"
  viewabilityConfigCallbackPairs={Array []}
>
  <View />
</RCTScrollView>
`;
